Forge.TreeWalker.UnitTests/test/TreeSchemaValidatorTests.cs

```diff
@@ -0,0 +1,208 @@
+            Assert.AreEqual(true, res);
+            Assert.AreEqual(0, errorList.Count);
+        }
+
+        [TestMethod]
+        public void Test_ValidateSchemasFromDirectoryListWithCustomRulesInJSchema()
+        {
+            bool res = ForgeSchemaValidator.ValidateSchemaFromDirectory(schemaDirectoryPath, jschemaRules, false, out IList<ValidationError> errorList);
+            Assert.AreEqual(true, res); 
+            Assert.AreEqual(0, errorList.Count);
+        }
+
+        [TestMethod]
+        public void Test_ValidateInvalidSchemaInString()
+        {
+            bool res = ForgeSchemaValidator.ValidateSchemaAsString(tardigradeSchemaString, jschemaRules, false, out IList<ValidationError> errorList);
+            Assert.AreEqual(true, res);
+            Assert.AreEqual(0, errorList.Count);
+        }
+
+        [TestMethod]
+        public void Test_ValidateInvalidSchemaDirectoryPath()
+        {
+            try
+            {
+                ForgeSchemaValidator.ValidateSchemaFromDirectory(invalidSchemaDirectoryPath, jschemaRules, false, out IList<ValidationError> errorList);
+            }
+            catch (Exception e)
+            {
+                Assert.AreEqual("The directory name is invalid.\r\n", e.Message);
+            }
+        }
+
+        [TestMethod]
+        public void Test_ValidateSchemaWithErrorContent()
+        {
+            bool res = ForgeSchemaValidator.ValidateSchemaAsString(invalidSchemaWithErrorContent, jschemaRules, false, out IList<ValidationError> errorList);
+            Assert.AreEqual(false, res);
+            Assert.AreEqual("JSON is valid against no schemas from 'oneOf'.", errorList.First().Message);
+        }
+
+        [TestMethod]
+        public void Test_ValidateInvalidSchemaAsDictionary()
+        {
+            bool res = ForgeSchemaValidator.ValidateSchemaAsString(invalidSchemaWithErrorContent, jschemaRules, true, out IList<ValidationError> errorList);
+            Assert.AreEqual(false, res);
+            Assert.AreEqual("JSON is valid against no schemas from 'oneOf'.", errorList.First().Message);
+        }
+
+        [TestMethod]
+        public void Test_ValidateValidSchemaAsDictionary()
+        {
+            bool res = ForgeSchemaValidator.ValidateSchemaAsString(subroutineSchemaString, rulesForDictionary, true, out IList<ValidationError> errorList);
+            Assert.AreEqual(true, res);
+            Assert.AreEqual(0, errorList.Count);
+        }
+
+        [TestMethod]
+        public void Test_ValidateAsDictionaryContainsDuplicateKey()
+        {
+            try
+            {
+                ForgeSchemaValidator.ValidateSchemaFromDirectory(schemaDirectoryPath, stringRules, true, out IList<ValidationError> errorList);
+            }
+            catch (Exception e)
+            {
+                Assert.AreEqual("An item with the same key has already been added.", e.Message);
+            }
```

Comment:
This test fails because you are validatingAsDictionary, but some of the schemas in the directory are not dictionaries. Also, you aren't using the dictionaryRules.

What you actually want to test here is that testing a directory that contains both ForgeTrees and Dictionaries, when validatingAsDictionary, should fail. This should fail because the ForgeTrees are not dictionaries.

This test didn't fail in that way because JSON actually deserialized the ForgeTrees into a dictionary, but the values are bogus. You should add a check to GetAllSchemasInDirectory to see if the forgeTree actually deserialized properly. If it doesn't, throw an exception.

I have the same check in the existing schema validation tests that you can use here. Check if forgeTree.Tree is null after deserializing as dictionary to confirm it is a dictionary. If forgeTree.Tree is null, that's when you should throw an exception.

Catch that exception in this UT.
Rename the UT:
Test_ValidateSchemaFromDirectory_WithValidateAsDictionary_DirectoryContainsForgeTrees_Fail